### PR TITLE
Testing Coprocessor Unusable and Reserved Instruction exception in userspace.

### PR DESCRIPTION
**Pull request by dingorth**
* testing Coprocessor Unusable exception in userspace.
* testing Reserved Instruction exception in userspace.

### CODE REVIEW

**Comment by dingorth** [comment]
How to raise Reserved Instruction exception in userspace? Does it really make sense to do that?

**Comment by dingorth** [comment]
Before merge we need to have Reserved Instruction exception handler implemented.

File: user/utest/misbehave.c [hunk 1]
```diff
@@ -24,3 +27,44 @@ int test_misbehave() {
...
 }
+
+static inline void exec_cp0_instr(void) {
+  int value;
+  asm volatile("mfc0 %0, $12, 0" : "=r"(value));
+}
+
+static inline void exec_reserved_instr(void) {
+  int value;
+  asm volatile("mfc2 %0, $12, 0" : "=r"(value));
```
Comment (cahirwpz) [inline user/utest/misbehave.c]: Some MIPS CPUs implement CP2 as mentioned on wikipedia:

> For example, in the PlayStation video game console, CP2 is the Geometry Transformation Engine (GTE), which accelerates the processing of geometry in 3D computer graphics.

Reserved instructions are according to my knowledge all privileged instructions such as `eret`, `tlbp` and so on. Refer to **Table 3.23 Privileged Instructions** from *MIPS® Architecture For Programmers Volume II-A: The MIPS32® Instruction Set*.

File: user/utest/misbehave.c [hunk 2]
```diff
@@ -24,3 +27,44 @@ int test_misbehave() {
...
+      exit(EXIT_FAILURE);
+    case 0:
+      proc_handler();
+      exit(EXIT_SUCCESS);
+    default:
+      return pid;
+  }
+}
+
+int test_userspace_cop_unusable(void) {
```
Comment (cahirwpz) [inline user/utest/misbehave.c]: Hence code for both functions is almost the same... please replace it with call to:
```c
static void check_exception(void (*code)(void), int signum)
```
... and move relevant code there.

File: user/utest/misbehave.c (hunk 1 above)
Comment (dingorth) [reply user/utest/misbehave.c]: Suprisingly all of the instructions from Table 3.23 cause Coprocessor Unusable exception. At least on ours Qemu setup.

File: user/utest/misbehave.c (hunk 2 above)
Comment (dingorth) [reply user/utest/misbehave.c]: Sure, I'll fix it.

File: user/utest/misbehave.c (hunk 1 above)
Comment (cahirwpz) [reply user/utest/misbehave.c]: Right! Another idea – refer to chapter A.2 of same document. Choose encoding that has `*` in it. According to *Table A.1 Symbols Used in the Instruction Encoding Tables*:

> Operation or field codes marked with this symbol are reserved for future use. Executing such an instruction must cause a Reserved Instruction Exception.

**Comment by dingorth** [comment]
Created issue #424

**Pull request merged**